packages/atomic-elements/src/components/Overlays/Modal/hooks/useManageModalScroll.ts

```diff
@@ -0,0 +1,41 @@
+import React, { useEffect, useRef } from "react";
+import { OverlayTriggerState } from "react-stately";
+
+interface UseManageModalScrollProps {
+  ensureVisible?: boolean;
+  modalRef: React.RefObject<HTMLElement | null>;
+  scrollOptions?: boolean | ScrollIntoViewOptions;
+}
+
+/**
+ * Custom hook that manages modal scrolling behavior and focus restoration.
+ *
+ * When the modal opens and `ensureVisible` is true, this hook will:
+ * - Store the currently focused element
+ * - Scroll the modal into view using the provided scroll options
+ *
+ * When the modal closes or the component unmounts, it will:
+ * - Restore focus to the previously focused element
+ * - Reset the stored focus reference
+ */
+export function useManageModalScroll(
+  props: UseManageModalScrollProps,
+  state: OverlayTriggerState
+) {
+  const { modalRef, ensureVisible, scrollOptions } = props;
+  const previousFocus = useRef<HTMLElement | null>(null);
+
+  useEffect(() => {
+    if (state.isOpen && ensureVisible) {
+      previousFocus.current = document.activeElement as HTMLElement | null;
+      modalRef.current?.scrollIntoView(scrollOptions);
+    }
+
+    return () => {
+      // Restore previous focus, which will move the page back to where it was.
+      // 99% of the time this will be the element that was focused before the modal opened.
+      previousFocus.current?.focus();
```

Comment:
Will this handle the case where clicking something in the modal removes that element, or navigates somewhere else?

Reply:
Hm, that's a good point. We should at least check if the element is still in the DOM before trying to focus it. If it's not we can probably just skip the focus? Not sure if there's anything we could do at that point

Reply:
Yeah, as long as we check it's still there I think that would be plenty. If it's not then there's not actually anything to do anyways I think, although we could maybe instead store the scroll position of the top of the view port prior to opening the modal and just scroll back to that? But you still run into the issue of if the modal navigated somewhere else then that might not make sense to do.